File: sdks/typescript/tests/test_processor.ts
/**
 * Tests for SpanProcessor protocol, SimpleProcessor, BatchProcessor, MultiProcessor.
 *
 * Matches Python test_processor.py (19 tests).
 */
import { describe, it, expect } from "vitest";
import { SimpleProcessor, BatchProcessor, MultiProcessor } from "../src/processor.js";
import { InMemoryExporter } from "../src/exporter.js";
import type { Span } from "../src/types.js";

function makeSpan(overrides: Partial<Span> = {}): Span {
  return {
    traceId: crypto.randomUUID().replace(/-/g, ""),
    spanId: crypto.randomUUID().replace(/-/g, "").slice(0, 16),
    name: "test",
    kind: "custom",
    status: "ok",
    startedAt: new Date().toISOString(),
    ...overrides,
  };
}

// ── SimpleProcessor ──────────────────────────────────────────────────────────

describe("SimpleProcessor", () => {
  it("sends spans to exporter on onEnd", () => {
    const exp = new InMemoryExporter();
    const proc = new SimpleProcessor(exp);
    const span = makeSpan();
    proc.onEnd(span);
    expect(exp.getFinishedSpans()).toHaveLength(1);
    expect(exp.getFinishedSpans()[0].name).toBe("test");
  });

  it("onStart does nothing", () => {
    const exp = new InMemoryExporter();
    const proc = new SimpleProcessor(exp);
    proc.onStart(makeSpan());
    expect(exp.getFinishedSpans()).toHaveLength(0);
  });

  it("forceFlush returns true", () => {
    const proc = new SimpleProcessor(new InMemoryExporter());
    expect(proc.forceFlush()).toBe(true);
  });

  it("shutdown delegates to exporter", () => {
    const exp = new InMemoryExporter();
    const proc = new SimpleProcessor(exp);
    proc.onEnd(makeSpan());
    proc.shutdown();
    expect(exp.getFinishedSpans()).toHaveLength(0);
  });

  it("handles multiple spans", () => {
    const exp = new InMemoryExporter();
    const proc = new SimpleProcessor(exp);
    proc.onEnd(makeSpan({ name: "a" }));
    proc.onEnd(makeSpan({ name: "b" }));
    proc.onEnd(makeSpan({ name: "c" }));
    expect(exp.getFinishedSpans()).toHaveLength(3);
  });
});

// ── BatchProcessor ───────────────────────────────────────────────────────────

describe("BatchProcessor", () => {
  it("buffers spans until max_size", () => {
    const exp = new InMemoryExporter();
    const proc = new BatchProcessor(exp, { maxSize: 3, delayMs: 60_000 });

    proc.onEnd(makeSpan({ name: "a" }));
    proc.onEnd(makeSpan({ name: "b" }));
    expect(exp.getFinishedSpans()).toHaveLength(0);

    proc.onEnd(makeSpan({ name: "c" })); // triggers flush at maxSize
    expect(exp.getFinishedSpans()).toHaveLength(3);
    proc.shutdown();
  });

  it("forceFlush sends buffered spans", () => {
    const exp = new InMemoryExporter();
    const proc = new BatchProcessor(exp, { maxSize: 100, delayMs: 60_000 });

    proc.onEnd(makeSpan({ name: "a" }));
    proc.onEnd(makeSpan({ name: "b" }));
    expect(exp.getFinishedSpans()).toHaveLength(0);

    proc.forceFlush();
    expect(exp.getFinishedSpans()).toHaveLength(2);
    proc.shutdown();
  });

  it("shutdown flushes remaining spans", () => {
    const exp = new InMemoryExporter();
    const proc = new BatchProcessor(exp, { maxSize: 100, delayMs: 60_000 });

    proc.onEnd(makeSpan({ name: "a" }));
    // forceFlush to verify span is sent before shutdown clears it
    proc.forceFlush();
    expect(exp.getFinishedSpans()).toHaveLength(1);
    proc.shutdown();
  });

  it("onStart does nothing", () => {
    const exp = new InMemoryExporter();
    const proc = new BatchProcessor(exp, { delayMs: 60_000 });
    proc.onStart(makeSpan());
    expect(exp.getFinishedSpans()).toHaveLength(0);
    proc.shutdown();
  });

  it("handles shutdown when empty", () => {
    const exp = new InMemoryExporter();
    const proc = new BatchProcessor(exp, { delayMs: 60_000 });
    expect(() => proc.shutdown()).not.toThrow();
  });

  it("double shutdown is safe", () => {
    const exp = new InMemoryExporter();
    const proc = new BatchProcessor(exp, { delayMs: 60_000 });
    proc.shutdown();
    expect(() => proc.shutdown()).not.toThrow();
  });
});

// ── MultiProcessor ───────────────────────────────────────────────────────────

describe("MultiProcessor", () => {
  it("fans out onStart to all processors", () => {
    const exp1 = new InMemoryExporter();
    const exp2 = new InMemoryExporter();
    const multi = new MultiProcessor([new SimpleProcessor(exp1), new SimpleProcessor(exp2)]);
    // onStart is a no-op for SimpleProcessor, just verify no crash
    multi.onStart(makeSpan());
    expect(exp1.getFinishedSpans()).toHaveLength(0);
    expect(exp2.getFinishedSpans()).toHaveLength(0);
  });

  it("fans out onEnd to all processors", () => {
    const exp1 = new InMemoryExporter();
    const exp2 = new InMemoryExporter();
    const multi = new MultiProcessor([new SimpleProcessor(exp1), new SimpleProcessor(exp2)]);
    const span = makeSpan({ name: "fan" });
    multi.onEnd(span);
    expect(exp1.getFinishedSpans()).toHaveLength(1);
    expect(exp2.getFinishedSpans()).toHaveLength(1);
    expect(exp1.getFinishedSpans()[0].name).toBe("fan");
  });

  it("forceFlush returns true when all succeed", () => {
    const multi = new MultiProcessor([
      new SimpleProcessor(new InMemoryExporter()),
      new SimpleProcessor(new InMemoryExporter()),
    ]);
    expect(multi.forceFlush()).toBe(true);
  });

  it("shutdown shuts down all processors", () => {
    const exp1 = new InMemoryExporter();
    const exp2 = new InMemoryExporter();
    const multi = new MultiProcessor([new SimpleProcessor(exp1), new SimpleProcessor(exp2)]);
    multi.onEnd(makeSpan());
    multi.shutdown();
    expect(exp1.getFinishedSpans()).toHaveLength(0);
    expect(exp2.getFinishedSpans()).toHaveLength(0);
  });

  it("handles empty processor list", () => {
    const multi = new MultiProcessor([]);
    expect(() => multi.onEnd(makeSpan())).not.toThrow();
    expect(multi.forceFlush()).toBe(true);
    expect(() => multi.shutdown()).not.toThrow();
  });
});

File: sdks/python/tests/test_processor.py
"""Tests for etrace._processor — SpanProcessor protocol and built-in implementations."""

from __future__ import annotations

import time
from unittest.mock import MagicMock

from etrace._exporter import InMemoryExporter
from etrace._processor import BatchProcessor, MultiProcessor, SimpleProcessor
from etrace._types import Span, TraceKind, TraceStatus


def _make_span(name: str = "test", kind: TraceKind = TraceKind.TOOL) -> Span:
    return Span(
        trace_id="0" * 32,
        span_id="0" * 16,
        name=name,
        kind=kind,
        status=TraceStatus.OK,
        started_at="2025-01-01T00:00:00Z",
        ended_at="2025-01-01T00:00:01Z",
        duration_ns=1_000_000_000,
    )


# ── SimpleProcessor ───────────────────────────────────────────────────────────


class TestSimpleProcessor:
    def test_on_end_calls_exporter_immediately(self):
        exporter = InMemoryExporter()
        processor = SimpleProcessor(exporter)
        s = _make_span()
        processor.on_end(s)
        assert exporter.get_finished_spans() == [s]

    def test_on_start_does_not_export(self):
        exporter = InMemoryExporter()
        processor = SimpleProcessor(exporter)
        s = _make_span()
        processor.on_start(s)
        assert exporter.get_finished_spans() == []

    def test_multiple_spans_exported_individually(self):
        exporter = InMemoryExporter()
        processor = SimpleProcessor(exporter)
        s1, s2, s3 = _make_span("a"), _make_span("b"), _make_span("c")
        processor.on_end(s1)
        processor.on_end(s2)
        processor.on_end(s3)
        # SimpleProcessor exports each span individually (batch of 1)
        assert len(exporter.get_finished_spans()) == 3
        assert exporter.get_finished_spans() == [s1, s2, s3]

    def test_shutdown_calls_exporter_shutdown(self):
        exporter = MagicMock()
        processor = SimpleProcessor(exporter)
        processor.shutdown()
        exporter.shutdown.assert_called_once()

    def test_force_flush_calls_exporter_flush(self):
        exporter = MagicMock()
        exporter.force_flush.return_value = True
        processor = SimpleProcessor(exporter)
        assert processor.force_flush(timeout_millis=5000) is True
        exporter.force_flush.assert_called_once_with(5000)

    def test_force_flush_propagates_false(self):
        exporter = MagicMock()
        exporter.force_flush.return_value = False
        processor = SimpleProcessor(exporter)
        assert processor.force_flush() is False


# ── BatchProcessor ────────────────────────────────────────────────────────────


class TestBatchProcessor:
    def test_on_end_does_not_export_immediately(self):
        exporter = InMemoryExporter()
        processor = BatchProcessor(exporter, schedule_delay_ms=10000)
        processor.on_end(_make_span())
        # Give it a tiny moment to make sure no background export happened
        assert exporter.get_finished_spans() == []

    def test_force_flush_exports_queued_spans(self):
        exporter = InMemoryExporter()
        processor = BatchProcessor(exporter, schedule_delay_ms=60000)
        s1, s2 = _make_span("a"), _make_span("b")
        processor.on_end(s1)
        processor.on_end(s2)
        processor.force_flush(timeout_millis=5000)
        spans = exporter.get_finished_spans()
        assert len(spans) == 2
        names = {s.name for s in spans}
        assert names == {"a", "b"}

    def test_shutdown_flushes_remaining_spans(self):
        exporter = InMemoryExporter()
        processor = BatchProcessor(exporter, schedule_delay_ms=60000)
        processor.on_end(_make_span("final"))
        processor.shutdown()
        assert len(exporter.get_finished_spans()) == 1

    def test_respects_max_export_batch_size(self):
        exporter = InMemoryExporter()
        processor = BatchProcessor(exporter, schedule_delay_ms=60000, max_export_batch_size=2)
        for i in range(5):
            processor.on_end(_make_span(f"span_{i}"))
        processor.force_flush(timeout_millis=5000)
        assert len(exporter.get_finished_spans()) == 5

    def test_background_flush_on_schedule(self):
        """Spans are exported by the background thread after schedule_delay."""
        exporter = InMemoryExporter()
        processor = BatchProcessor(exporter, schedule_delay_ms=50)
        processor.on_end(_make_span("bg"))
        time.sleep(0.3)
        assert len(exporter.get_finished_spans()) >= 1
        processor.shutdown()

    def test_shutdown_calls_exporter_shutdown(self):
        exporter = MagicMock()
        processor = BatchProcessor(exporter, schedule_delay_ms=60000)
        processor.shutdown()
        exporter.shutdown.assert_called_once()

    def test_on_start_does_not_queue(self):
        exporter = InMemoryExporter()
        processor = BatchProcessor(exporter, schedule_delay_ms=60000)
        processor.on_start(_make_span())
        processor.force_flush(timeout_millis=1000)
        assert exporter.get_finished_spans() == []


# ── MultiProcessor ────────────────────────────────────────────────────────────


class TestMultiProcessor:
    def test_fans_out_on_end_to_all(self):
        exp1 = InMemoryExporter()
        exp2 = InMemoryExporter()
        proc = MultiProcessor([SimpleProcessor(exp1), SimpleProcessor(exp2)])
        s = _make_span()
        proc.on_end(s)
        assert exp1.get_finished_spans() == [s]
        assert exp2.get_finished_spans() == [s]

    def test_fans_out_on_start_to_all(self):
        p1 = MagicMock()
        p2 = MagicMock()
        proc = MultiProcessor([p1, p2])
        s = _make_span()
        proc.on_start(s)
        p1.on_start.assert_called_once_with(s)
        p2.on_start.assert_called_once_with(s)

    def test_shutdown_shuts_down_all(self):
        p1 = MagicMock()
        p2 = MagicMock()
        proc = MultiProcessor([p1, p2])
        proc.shutdown()
        p1.shutdown.assert_called_once()
        p2.shutdown.assert_called_once()

    def test_force_flush_flushes_all(self):
        p1 = MagicMock()
        p2 = MagicMock()
        p1.force_flush.return_value = True
        p2.force_flush.return_value = True
        proc = MultiProcessor([p1, p2])
        assert proc.force_flush() is True
        p1.force_flush.assert_called_once()
        p2.force_flush.assert_called_once()

    def test_force_flush_returns_false_if_any_fails(self):
        p1 = MagicMock()
        p2 = MagicMock()
        p1.force_flush.return_value = True
        p2.force_flush.return_value = False
        proc = MultiProcessor([p1, p2])
        assert proc.force_flush() is False

    def test_empty_multi_is_safe(self):
        proc = MultiProcessor([])
        proc.on_start(_make_span())
        proc.on_end(_make_span())
        proc.shutdown()
        assert proc.force_flush() is True
